File: 1057_campus_bikes.py
from typing import List
from collections import defaultdict


class Solution:
    def assignBikes(self, workers: List[List[int]], bikes: List[List[int]]) -> List[int]:
        """
            https://leetcode.com/problems/campus-bikes/
            Time Complexity - O(mn + logk)
            m - number of workers
            n - number of bikes
            k - number of distances in the dictionary
            Space Complexity - O(mn)
        """
        # edge case
        if not workers or not bikes:
            return []

        dic = defaultdict(list)
        for w_idx, worker in enumerate(workers):
            for b_idx, bike in enumerate(bikes):
                dist = self._calculate_dist(worker, bike)
                dic[dist].append((w_idx, b_idx))

        result = [0] * len(workers)
        workers_assigned = [False] * len(workers)
        bikes_assigned = [False] * len(bikes)
        for dist in sorted(dic.keys()):
            for worker, bike in dic[dist]:
                if not workers_assigned[worker] and not bikes_assigned[bike]:
                    workers_assigned[worker] = True
                    bikes_assigned[bike] = True
                    result[worker] = bike
        return result

    def _calculate_dist(self, worker, bike):
        return abs(worker[0] - bike[0]) + abs(worker[1] - bike[1])


if __name__ == '__main__':
    print(Solution().assignBikes([[0, 0], [1, 1], [2, 0]], [[1, 0], [2, 2], [2, 1]]))

File: 1057. Campus Bikes.java
//1057. Campus Bikes
//Time Complexity:mnlogmn
//Space Complexity:mn

class Solution {
    public int[] assignBikes(int[][] workers, int[][] bikes) {
		
		//create a treemap so we can access it in the sorted manner

		int m  = workers.length;
		int n  = bikes.length;  
        
        Map<Integer,List<int[]>> map = new TreeMap<>();
        //create map
        for(int i = 0; i < m; ++i){
            
            for(int j = 0; j < n; ++j){
                int dist = calculateDist(workers[i],bikes[j]);        
                
                //check if this dist already in the map 
                //else initialize the list
                if(!map.containsKey(dist)){
                    map.put(dist, new ArrayList<int[]>());
                }
                List<int[]> tempList = map.get(dist);
                tempList.add(new int[] {i,j});
                map.put(dist,tempList);
            }
        }
    
    int result[] = new int[m];
    HashSet<Integer> workersDone = new HashSet();
    HashSet<Integer> bikesDone = new HashSet();
    //done with the creation of the hashmap
    //iterate through the hashmap and perform the greedy choice
    for(int key: map.keySet()){
        //for the given distance go through all the pairs
        for( int[] pair: map.get(key)){
            
            int worker = pair[0];
            int bike = pair[1];
            if(!bikesDone.contains(bike) && !workersDone.contains(worker)){
                //if non of them are processed/used yet
                result[worker] = bike;
                workersDone.add(worker);
                bikesDone.add(bike);
            }   
            
        }
        
    }
    return result;
    }
    
    private int calculateDist(int[] worker, int[] bike){
        return (Math.abs(worker[0] - bike[0]) + Math.abs(worker[1] - bike[1]));
    }
}
